code/datums/components/stun_mitigation.dm

```diff
@@ -0,0 +1,186 @@
+/datum/component/stun_mitigation
+	var/mob/living/affected
+	///Percentage chance to mitigate a hardstun
+	var/datum/armor/cover
+	///Any special behavior flags
+	var/shield_flags = NONE
+	///The slots in which the parent can be in for the component to apply
+	var/slot_flags = SLOT_L_HAND|SLOT_R_HAND
+	///Whether the component is currently active
+	var/active = TRUE
+
+
+/datum/component/stun_mitigation/Initialize(shield_flags, slot_override, shield_cover = list(MELEE = 100, BULLET = 100, LASER = 100, ENERGY = 100, BOMB = 100, BIO = 100, FIRE = 100, ACID = 100))
+	. = ..()
+	if(!isitem(parent))
+		return COMPONENT_INCOMPATIBLE
+
+	var/obj/item/parent_item = parent
+
+	if(slot_override)
+		slot_flags = slot_override
+
+	if(shield_flags & SHIELD_TOGGLE)
+		RegisterSignal(parent, COMSIG_ITEM_TOGGLE_ACTIVE, PROC_REF(toggle_shield))
+		active = parent_item.active
+
+	RegisterSignal(parent, COMSIG_ITEM_EQUIPPED, PROC_REF(shield_equipped))
+	RegisterSignal(parent, COMSIG_ITEM_DROPPED, PROC_REF(shield_dropped))
+
+	if(active)
+		if(ismob(parent_item.loc))
+			shield_init(parent_item.loc)
+
+	if(islist(shield_cover))
+		cover = getArmor(arglist(shield_cover))
+	else if(istype(shield_cover, /datum/armor))
+		cover = shield_cover
+	else
+		cover = getArmor()
+		stack_trace("Invalid type found in cover during /datum/component/stun_mitigation Initialize()")
+
+
+/datum/component/stun_mitigation/Destroy()
+	shield_detatch_from_user()
+	cover = null
+	return ..()
+
+///A horrible proc for if the shield is initialised in a slot
+/datum/component/stun_mitigation/proc/shield_init(mob/holder_mob)
```

Comment:
I kind of hate this. When are shield comps getting created after an item is already equipped?

Reply:
I did wonder this myself, but the shield component has it (and basically everything that has this component also has the shield one) so I assume it can happen somehow. I didn't actually check though, so maybe not?